src/com/xilinx/rapidwright/edif/ParallelEDIFParser.java

```diff
@@ -0,0 +1,283 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.xilinx.rapidwright.device.Device;
+import com.xilinx.rapidwright.tests.CodePerfTracker;
+import com.xilinx.rapidwright.util.ParallelismTools;
+import com.xilinx.rapidwright.util.function.InputStreamSupplier;
+
+/**
+ * Fast EDIF Parser using parallelism
+ */
+public class ParallelEDIFParser implements AutoCloseable{
+    private static final long SIZE_PER_THREAD = EDIFTokenizer.DEFAULT_MAX_TOKEN_LENGTH * 8L;
+    protected final List<ParallelEDIFParserWorker> workers = new ArrayList<>();
+    protected final Path fileName;
+    private final long fileSize;
+    protected final InputStreamSupplier inputStreamSupplier;
+    protected final int maxTokenLength;
+    protected NameUniquifier uniquifier = NameUniquifier.concurrentUniquifier();
+
+    ParallelEDIFParser(Path fileName, long fileSize, InputStreamSupplier inputStreamSupplier, int maxTokenLength) {
+        this.fileName = fileName;
+        this.fileSize = fileSize;
+        this.inputStreamSupplier = inputStreamSupplier;
+        this.maxTokenLength = maxTokenLength;
+    }
+
+    public ParallelEDIFParser(Path fileName, long fileSize, InputStreamSupplier inputStreamSupplier) {
+        this(fileName, fileSize, inputStreamSupplier, EDIFTokenizer.DEFAULT_MAX_TOKEN_LENGTH);
+    }
+
+    public ParallelEDIFParser(Path p, long fileSize) {
+        this(p, fileSize, InputStreamSupplier.fromPath(p));
+    }
+
+    public ParallelEDIFParser(Path p) throws IOException {
+        this(p, Files.size(p));
+    }
+
+    protected ParallelEDIFParserWorker makeWorker(long offset) throws IOException {
+        return new ParallelEDIFParserWorker(fileName, inputStreamSupplier.get(), offset, uniquifier, maxTokenLength);
+    }
+
+    private int calcThreads(long fileSize) {
+        int maxUsefulThreads = Math.max((int) (fileSize / SIZE_PER_THREAD),1);
+        int processors = Runtime.getRuntime().availableProcessors();
+        return Math.min(maxUsefulThreads, processors);
+    }
+
+
+    protected void initializeWorkers() throws IOException {
+        workers.clear();
+        int threads = calcThreads(fileSize);
+        long offsetPerThread = fileSize / threads;
+        for (int i=0;i<threads;i++) {
+            ParallelEDIFParserWorker worker = makeWorker(i*offsetPerThread);
+            workers.add(worker);
+        }
+    }
+
+    private int numberOfThreads;
+    public EDIFNetlist parseEDIFNetlist(CodePerfTracker t) throws IOException {
+
+        t.start("Initialize workers");
+        initializeWorkers();
+        numberOfThreads = workers.size();
+
+        t.stop().start("Parse First Token");
+        final List<ParallelEDIFParserWorker> failedWorkers = ParallelismTools.maybeToParallel(workers.stream())
```

Comment:
Stream parallelism uses `ForkJoinPool.commonPool()` but I would prefer parallel activity to be funneled through `ParallelismTools` (via `invokeAll`) and use the thread pool it has. This way, we can fully control current and future parallel activity ourselves, and this could possibly lead to a more scalable solution because it would support the case where number of workers/tasks exceeds the number of processors in order to deal with work imbalance arising from EDIF parsing or other parsing activities.

What do you think?

Reply:
`ParallelismTools` was built to be a *parallel* (hee hee) to `ForkJoinPool`, which also has one thread less than the number of procs. The reason for this design is that, for blocking operations (i.e. an `invoke` which blocks until all jobs are done) the blocking thread will also "steal" tasks from the queue. In non-blocking operations, (i.e. a `submit`) then the submitting thread can continue and do some other async work without causing the number of working threads to exceed the number of processors.

Where it differs from the `ForkJoinPool` is that it has a shared thread queue (rather than one queue per thread) which I've found to have give better thread utilization for coarse-grained tasks, rather than the finer-grained divide-and-conquer tasks generated by `parallelStream()`.

I'm not fully convinced that what I've built is indeed the best thing, so if you do dive into it please critique!